src/Symfony/Component/HttpClient/NoPrivateNetworkHttpClient.php

```diff
@@ -0,0 +1,128 @@
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Psr\Log\LoggerAwareInterface;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpClient\Exception\InvalidArgumentException;
+use Symfony\Component\HttpClient\Exception\TransportException;
+use Symfony\Component\HttpFoundation\IpUtils;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+
+/**
+ * Decorator that block requests to private network by default.
+ *
+ * @author Hallison Boaventura <hallisonboaventura@gmail.com>
+ */
+final class NoPrivateNetworkHttpClient implements HttpClientInterface, LoggerAwareInterface
+{
+    use HttpClientTrait;
+
+    const IPV4_PRIVATE_SUBNETS = [
+        '127.0.0.0/8',
+        '10.0.0.0/8',
+        '192.168.0.0/16',
+        '172.16.0.0/12',
+        '169.254.0.0/16',
+        '0.0.0.0/8',
+        '240.0.0.0/4',
+    ];
+
+    const IPV6_PRIVATE_SUBNETS = [
+        '::1/128',
+        'fc00::/7',
+        'fe80::/10',
+        '::ffff:0:0/96',
+        '::/128',
+    ];
+
+    /**
+     * @var HttpClientInterface
+     */
+    private $client;
+
+    /**
+     * Subnets in CIDR notation that IpUtils will make use of.
+     *
+     * @var string|array|null
+     */
+    private $subnets;
+
+    /**
+     * Constructor.
+     *
+     * @param HttpClientInterface $client  A HttpClientInterface instance.
+     * @param string|array|null   $subnets String or array of subnets using CIDR notation that will be used by IpUtils.
+     *                                     If its value is null, then default private subnets will be used.
+     */
+    public function __construct(HttpClientInterface $client, $subnets = null)
+    {
+        $this->client = $client;
+
+        if (!(\is_string($subnets) || \is_array($subnets) || null === $subnets)) {
+            throw new InvalidArgumentException('$subnets argument must be string, array or null.');
+        }
+
+        $this->subnets = $subnets;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        $subnets = $this->subnets;
+
+        $lastPrimaryIp = '';
+        $onProgress = $options['on_progress'] ?? null;
+
+        $options['on_progress'] = function (int $dlNow, int $dlSize, array $info) use ($onProgress, $subnets, &$lastPrimaryIp): void {
+            if ($info['primary_ip'] !== $lastPrimaryIp) {
+                if (
+                    (null !== $subnets && IpUtils::checkIp($info['primary_ip'], $subnets)) ||
+                    (null === $subnets && (
+                        (0 < substr_count($info['primary_ip'], '.') && IpUtils::checkIp($info['primary_ip'], self::IPV4_PRIVATE_SUBNETS)) ||
+                        (0 < substr_count($info['primary_ip'], ':') && IpUtils::checkIp($info['primary_ip'], self::IPV6_PRIVATE_SUBNETS))
```

Comment:
shouldn't we merge both IPV4/V6 constants in one since `IpUtils::checkIp()` is able to deal with both?
this would reduce the logic to
`IpUtils::checkIp($info['primary_ip'], $subnets ?? self::PRIVATE_SUBNETS)`

Reply:
Implementation will be more simpler, but we'll lose performance by checking IPv4 on IPv6 subnets (that will return false, which is expected), but still are unnecessary checks, is that OK?

Reply:
Yes, I think that's OK, the check is fast compared to the network anyway.